# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/ClientWebGLContext.cpp
# Commit: 0b191b707750
# Full Hash: 0b191b707750e040289033359306e4966c1bf062
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-06-15 21:48:38
# Regressor Bug: 1632249
# File Overlap Count: 1
# Description:
#   Bug 1632249 - WebXR compositing fixes. r=imanol
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D78117
# ==============================================================================

diff -r 16c5dd2cfd60 -r 0b191b707750 dom/canvas/ClientWebGLContext.cpp
--- a/dom/canvas/ClientWebGLContext.cpp	Mon Jun 15 18:25:55 2020 +0000
+++ b/dom/canvas/ClientWebGLContext.cpp	Mon Jun 15 18:25:58 2020 +0000
@@ -398,7 +398,17 @@
 
 // ------------------------- Composition, etc -------------------------
 
-void ClientWebGLContext::OnBeforePaintTransaction() { Present(); }
+void ClientWebGLContext::OnBeforePaintTransaction() {
+  const RefPtr<layers::ImageBridgeChild> imageBridge =
+      layers::ImageBridgeChild::GetSingleton();
+
+  auto texType = layers::TextureType::Unknown;
+  if (imageBridge) {
+    texType = layers::PreferredCanvasTextureType(*imageBridge);
+  }
+
+  Present(nullptr, texType);
+}
 
 void ClientWebGLContext::EndComposition() {
   // Mark ourselves as no longer invalidated.
@@ -407,43 +417,18 @@
 
 // -
 
-namespace webgl {
-
-void Present(ClientWebGLContext& webgl) { webgl.Present(); }
-
-Maybe<layers::SurfaceDescriptor> GetFrontBuffer(
-    ClientWebGLContext& webgl, layers::KnowsCompositor* const kc) {
-  auto texType = layers::TextureType::Unknown;
-  if (kc) {
-    texType = layers::PreferredCanvasTextureType(*kc);
-  }
-  return webgl.GetFrontBuffer(texType);
-}
-
-}  // namespace webgl
-
-// -
-
-void ClientWebGLContext::Present() {
+void ClientWebGLContext::Present(WebGLFramebufferJS* const fb,
+                                 const layers::TextureType type) {
   if (!mIsCanvasDirty) return;
   mIsCanvasDirty = false;
-  mFrontBufferDesc = nullptr;
   mFrontBufferSnapshot = nullptr;
 
-  Run<RPROC(Present)>();
+  Run<RPROC(Present)>(fb ? fb->mId : 0, type);
 }
 
 Maybe<layers::SurfaceDescriptor> ClientWebGLContext::GetFrontBuffer(
-    const layers::TextureType type) {
-  if (!mFrontBufferDesc) {
-    const auto desc = Run<RPROC(GetFrontBuffer)>(type);
-    if (desc) {
-      mFrontBufferDesc = MakeUnique<layers::SurfaceDescriptor>(*desc);
-    }
-  }
-
-  if (!mFrontBufferDesc) return Nothing();
-  return Some(*mFrontBufferDesc);
+    WebGLFramebufferJS* const fb, const layers::TextureType type) {
+  return Run<RPROC(GetFrontBuffer)>(fb ? fb->mId : 0, type);
 }
 
 // -
@@ -620,9 +605,6 @@
     }
     if (size == curSize) return NS_OK;  // MUST skip no-op resize
 
-    if (mIsCanvasDirty) {
-      Present();
-    }
     state.mDrawingBufferSize = Nothing();
     Run<RPROC(Resize)>(size);
 